osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -184,24 +184,8 @@ private void handleControlPoints(TextWriter writer)
             // iterate over hitobjects and pull out all required sample and difficulty changes
             foreach (var h in beatmap.HitObjects)
             {
-                if (isOsuRuleset)
-                {
-                    var hDifficultyPoint = h.DifficultyControlPoint;
-
-                    if (!hDifficultyPoint.IsRedundant(lastRelevantDifficultyPoint))
-                    {
-                        legacyControlPoints.Add(hDifficultyPoint.Time, hDifficultyPoint);
-                        lastRelevantDifficultyPoint = hDifficultyPoint;
-                    }
-                }
-
-                var hSamplePoint = h.SampleControlPoint;
-
-                if (!hSamplePoint.IsRedundant(lastRelevantSamplePoint))
-                {
-                    legacyControlPoints.Add(hSamplePoint.Time, hSamplePoint);
-                    lastRelevantSamplePoint = hSamplePoint;
-                }
+                extractDifficultyControlPoints(h);
+                extractSampleControlPoints(h);
```

Comment:
Maybe not super relevant, but I'll mention it: the way these are implemented sort of breaks down with concurrent objects. Consider something like the following case:

```
-------------------------------------> time

 [x]-----[o]-----[o]-----[o]-----[o]
     [x]
             [x]
                     [x]
```

where `[x]` signifies a hitobject's start time, and `[o]` signified a nested's start time. This procedure as written is silently relying on the fact that there are no other simultaneous hitobjects while an object with nested is being processed.

For full correctness you'd probably have to collect all hitobjects to a flat list, sort by start time and only then process. (This still doesn't fully solve a potential issue with hitobjects with duration, wherein you may have to "restore" the previous control point after processing another concurrent one, but that may be out of scope for now.)

This will only be relevant when encoding a beatmap with per-object points back into the legacy format. A new format that doesn't attempt to serialise back to the old format will not have such issues.